### ISSUE TITLE
firebase_database@0.1.3 causes a crash when disposed

### ISSUE THREAD

**Issue by eliezedeck** [flutter/flutter#13295]
## Steps to Reproduce

Use https://github.com/flutter/friendlychat-steps/tree/master/full_steps/step_5_data_syncing as the base, and upgrade all packages to latest version:

```yaml
dependencies:
  flutter:
    sdk: flutter

  # The following adds the Cupertino Icons font to your application.
  # Use with the CupertinoIcons class for iOS style icons.
  cupertino_icons: ^0.1.0

  google_sign_in: 1.0.1
  firebase_analytics: 0.1.1
  firebase_auth: 0.3.1
  firebase_database: 0.1.3
```

Create 2 screens, and use `Navigator` to go the `ChatScreen`. Then go back to the first screen so that the `ChatScreen` will be cleaned-up.

## Logs

When the `ChatScreen` is destroyed, this is the logs:

```text
I/flutter ( 6655): ══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
I/flutter ( 6655): The following UnsupportedError was thrown while finalizing the widget tree:
I/flutter ( 6655): Unsupported operation: List cannot be modified.
I/flutter ( 6655): When the exception was thrown, this was the stack:
I/flutter ( 6655): #0      FirebaseSortedList.length= (package:firebase_database/ui/firebase_sorted_list.dart:64)
I/flutter ( 6655): #1      Object&ListMixin.clear (dart:collection/list.dart:307)
I/flutter ( 6655): #2      FirebaseAnimatedListState.dispose (package:firebase_database/ui/firebase_animated_list.dart:168)
I/flutter ( 6655): #3      StatefulElement.unmount (package:flutter/src/widgets/framework.dart:3709)
I/flutter ( 6655): #4      _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:1730)
I/flutter ( 6655): #5      _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:1728)
I/flutter ( 6655): #6      ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:3586)
I/flutter ( 6655): #7      _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:1726)
I/flutter ( 6655): #8      _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:1728)
I/flutter ( 6655): #9      MultiChildRenderObjectElement.visitChildren (package:flutter/src/widgets/framework.dart:4621)
I/flutter ( 6655): #10     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:1726)
I/flutter ( 6655): #11     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:1728)
I/flutter ( 6655): #12     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:3586)
I/flutter ( 6655): #13     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:1726)
I/flutter ( 6655): #14     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:1728)
I/flutter ( 6655): #15     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:3586)
I/flutter ( 6655): #16     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:1726)
I/flutter ( 6655): #17     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:1728)
I/flutter ( 6655): #18     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:3586)
I/flutter ( 6655): #19     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:1726)
I/flutter ( 6655): #20     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:1728)
I/flutter ( 6655): #21     MultiChildRenderObjectElement.visitChildren (package:flutter/src/widgets/framework.dart:4621)
I/flutter ( 6655): #22     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:1726)
I/flutter ( 6655): #23     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:1728)
I/flutter ( 6655): #24     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:3586)
I/flutter ( 6655): #25     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:1726)
I/flutter ( 6655): #26     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:1728)
I/flutter ( 6655): #27     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:3586)
I/flutter ( 6655): #28     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:1726)
I/flutter ( 6655): #29     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:1728)
I/flutter ( 6655): #30     SingleChildRenderObjectElement.visitChildren (package:flutter/src/widgets/framework.dart:4521)
I/flutter ( 6655): #31     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:1726)
I/flutter ( 6655): #32     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:1728)
I/flutter ( 6655): #33     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:3586)
I/flutter ( 6655): #34     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:1726)
I/flutter ( 6655): #35     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:1728)
I/flutter ( 6655): #36     SingleChildRenderObjectElement.visitChildren (package:flutter/src/widgets/framework.dart:4521)
I/flutter ( 6655): #37     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:1726)
I/flutter ( 6655): #38     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:1728)
I/flutter ( 6655): #39     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:3586)
I/flutter ( 6655): #40     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:1726)
I/flutter ( 6655): #41     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:1728)
I/flutter ( 6655): #42     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:3586)
I/flutter ( 6655): #43     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:1726)
I/flutter ( 6655): #44     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:1728)
I/flutter ( 6655): #45     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:3586)
I/flutter ( 6655): #46     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:1726)
I/flutter ( 6655): #47     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:1728)
I/flutter ( 6655): #48     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:3586)
I/flutter ( 6655): #49     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:1726)
I/flutter ( 6655): #50     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:1728)
I/flutter ( 6655): #51     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:3586)
I/flutter ( 6655): #52     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:1726)
I/flutter ( 6655): #53     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:1728)
I/flutter ( 6655): #54     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:3586)
I/flutter ( 6655): #55     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:1726)
I/flutter ( 6655): #56     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:1728)
I/flutter ( 6655): #57     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:3586)
I/flutter ( 6655): #58     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:1726)
I/flutter ( 6655): #59     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:1728)
I/flutter ( 6655): #60     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:3586)
I/flutter ( 6655): #61     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:1726)
I/flutter ( 6655): #62     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:1728)
I/flutter ( 6655): #63     SingleChildRenderObjectElement.visitChildren (package:flutter/src/widgets/framework.dart:4521)
I/flutter ( 6655): #64     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:1726)
I/flutter ( 6655): #65     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:1728)
I/flutter ( 6655): #66     SingleChildRenderObjectElement.visitChildren (package:flutter/src/widgets/framework.dart:4521)
I/flutter ( 6655): #67     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:1726)
I/flutter ( 6655): #68     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:1728)
I/flutter ( 6655): #69     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:3586)
I/flutter ( 6655): #70     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:1726)
I/flutter ( 6655): #71     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:1728)
I/flutter ( 6655): #72     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:3586)
I/flutter ( 6655): #73     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:1726)
I/flutter ( 6655): #74     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:1728)
I/flutter ( 6655): #75     SingleChildRenderObjectElement.visitChildren (package:flutter/src/widgets/framework.dart:4521)
I/flutter ( 6655): #76     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:1726)
I/flutter ( 6655): #77     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:1728)
I/flutter ( 6655): #78     SingleChildRenderObjectElement.visitChildren (package:flutter/src/widgets/framework.dart:4521)
I/flutter ( 6655): #79     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:1726)
I/flutter ( 6655): #80     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:1728)
I/flutter ( 6655): #81     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:3586)
I/flutter ( 6655): #82     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:1726)
I/flutter ( 6655): #83     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:1728)
I/flutter ( 6655): #84     SingleChildRenderObjectElement.visitChildren (package:flutter/src/widgets/framework.dart:4521)
I/flutter ( 6655): #85     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:1726)
I/flutter ( 6655): #86     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:1728)
I/flutter ( 6655): #87     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:3586)
I/flutter ( 6655): #88     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:1726)
I/flutter ( 6655): #89     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:1728)
I/flutter ( 6655): #90     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:3586)
I/flutter ( 6655): #91     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:1726)
I/flutter ( 6655): #92     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:1728)
I/flutter ( 6655): #93     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:3586)
I/flutter ( 6655): #94     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:1726)
I/flutter ( 6655): #95     _InactiveElements._unmountAll (package:flutter/src/widgets/framework.dart:1740)
I/flutter ( 6655): #96     BuildOwner.finalizeTree.<anonymous closure> (package:flutter/src/widgets/framework.dart:2337)
I/flutter ( 6655): #97     BuildOwner.lockState (package:flutter/src/widgets/framework.dart:2169)
I/flutter ( 6655): #98     BuildOwner.finalizeTree (package:flutter/src/widgets/framework.dart:2336)
I/flutter ( 6655): #99     BindingBase&SchedulerBinding&GestureBinding&ServicesBinding&RendererBinding&WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:582)
I/flutter ( 6655): #100    BindingBase&SchedulerBinding&GestureBinding&ServicesBinding&RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:200)
I/flutter ( 6655): #101    BindingBase&SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:713)
I/flutter ( 6655): #102    BindingBase&SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:649)
I/flutter ( 6655): #103    _invoke (file:///b/build/slave/Linux_Engine/build/src/flutter/lib/ui/hooks.dart:91)
I/flutter ( 6655): #104    _drawFrame (file:///b/build/slave/Linux_Engine/build/src/flutter/lib/ui/hooks.dart:80)
I/flutter ( 6655): ════════════════════════════════════════════════════════════════════════════════════════════════════
```

## Flutter Doctor

```text
[✓] Flutter (on Mac OS X 10.12.6 16G1036, locale en-US, channel alpha)
    • Flutter at /Users/elie/ProjectsRuntimes/flutter-sdk
    • Framework revision e8aa40eddd (6 weeks ago), 2017-10-17 15:42:40 -0700
    • Engine revision 7c4142808c
    • Tools Dart version 1.25.0-dev.11.0

[✓] Android toolchain - develop for Android devices (Android SDK 27.0.0)
    • Android SDK at /Users/elie/Library/Android/sdk
    • Platform android-27, build-tools 27.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-915-b08)

[✓] iOS toolchain - develop for iOS devices (Xcode 9.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 9.1, Build version 9B55
    • ios-deploy 1.9.2
    • CocoaPods version 1.2.0

[✓] Android Studio (version 3.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-915-b08)

[✓] IntelliJ IDEA Ultimate Edition (version 2017.2.6)
    • Flutter plugin version 19.1
    • Dart plugin version 172.4343.25

[✓] Connected devices
    • Android SDK built for x86 • emulator-5554    • android-x86 • Android 6.0 (API 23) (emulator)
```

## Negative impacts on UI

Obviously, you can't really deal with it cleanly. Secondly, on an old phone, the *UI will freeze* for about 1 - 1.5 seconds. On the emulator, the `ChatScreen` page will start to gradually go down normally (animation), then about 1/3-way, it will abruptly disappear.

**Comment by collinjackson** [comment]
Looks like this was fixed in flutter/plugins#288

**Comment by kroikie** [comment]
@eliezedeck

This issue has been moved to https://github.com/FirebaseExtended/flutterfire/issues/932. Any further collaboration will be done there.
